File: Website/ClientApp/expression-tree/SqlExpression.ts
import { SqlExpressionType } from "./SqlExpressionType";
import { SqlType } from "./SqlType";
import { TransferObjectId } from "./TransferObjectId";
import { IExpressionTransferObject } from "./IExpressionTransferObject";

export interface SqlExpression extends IExpressionTransferObject {
    ExpressionType: SqlExpressionType;
    ReturnType: SqlType;
}

export class SqlExpressionBuilder {
    constructor() {
    }

    protected ExpressionType: SqlExpressionType;
    protected ReturnType: SqlType;


    setExpressionType(value: SqlExpressionType): SqlExpressionBuilder {
        this.ExpressionType = value;
        return this;
    }

    setReturnType(value: SqlType): SqlExpressionBuilder {
        this.ReturnType = value;
        return this;
    }

    build(): SqlExpression {
        return {
            ExpressionType: this.ExpressionType,
            ReturnType: this.ReturnType,
            TransferObjectId: TransferObjectId.SqlQueryExpression // Tu coś innego trzeba ustawić...
        };
    }
}

File: SqlExpressions/Expressions/SqlExpression.cs
using System;

namespace SqlExpressions.Expressions
{

    /// <summary>
    /// Base class for all types representing expressions.
    /// </summary>
    public abstract class SqlExpression : IExpressionTransferObject
    {
        /// <summary>
        /// Gets the return type of the expression.
        /// </summary>
        /// <value>
        /// The type of the return.
        /// </value>
        public abstract SqlType ReturnType { get; }
        /// <summary>
        /// Gets the type of the expression.
        /// </summary>
        /// <value>
        /// The type of the expression.
        /// </value>
        public abstract SqlExpressionType ExpressionType { get; }

        /// <summary>
        /// Gets the transfer object identifier.
        /// </summary>
        /// <value>
        /// The transfer object identifier.
        /// </value>
        public TransferObjectId TransferObjectId 
            => (TransferObjectId)(Enum.Parse(typeof(TransferObjectId), this.GetType().Name));

        /// <summary>
        /// Accepts the specified visitor.
        /// </summary>
        /// <typeparam name="T">Return type</typeparam>
        /// <param name="visitor">The visitor.</param>
        /// <returns>Returns processed value from expression</returns>
        public abstract T Accept<T>(ISqlVisitor<T> visitor);
    }
}
